### PR TITLE
First prototype [Work In Progress]

### PR DESCRIPTION
**Pull request by rivera-ernesto**
A first try to implement #1:
- A minimal `SLLog` object that keeps track of a `loggerClass` and some common _marker_ logic.
- A `SLLogger` protocol for log framework adapters.
- A `SLMarker` protocol and a `SLSimpleMarker` implementation of that protocol to support context-aware log frameworks.
- Simple `SLMacros` and `SLLibMacros` to be used by Apps and 3rd party libraries respectively.

The façade is "one way" only:
- Wrap log calls to direct them to log frameworks
- ~~Wrap log frameworks configuration/usage.~~

### CODE REVIEW

File: Source/SLLibMacros.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+//  LoggingDemo
+//
+//  Created by 利辺羅 on 2014/12/10.
+//  Copyright (c) 2014年 SLF4C. All rights reserved.
+//
+
+#import "SLMacros.h"
+
+#undef  SLMarkerName
+#define SLMarkerName @"Other"
```
Comment (bpoplauschi) [inline Source/SLLibMacros.h]: Let me get this straight: the idea is the default logging will use `'Other'` unless the SLMarkerName is redefined. This sounds right to me

File: Source/SLLogger.h [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+typedef NS_ENUM(NSUInteger, SLLevel)
+{
+    SLLevelOff = 0,
+    SLLevelVerbose,
+    SLLevelDebug,
+    SLLevelInfo,
+    SLLevelWarning,
+    SLLevelError
+};
```
Comment (bpoplauschi) [inline Source/SLLogger.h]: About future extensibility, couldn't we use a log levels configuration file (i.e. plist)? That way, regular users will have a prebuilt version of the file and the others will be able to change it without changing the actual code.

File: Source/SLLogger.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+
+@interface SLLog : NSObject
+
++ (Class<SLLogger>)loggerClass;
++ (void)setLoggerClass:(Class<SLLogger>)loggerClass;
+
++ (id<SLMarker>)appMarker;
++ (NSArray *)allMarkers;
++ (id<SLMarker>)markerWithName:(NSString *)name;
```
Comment (bpoplauschi) [inline Source/SLLogger.h]: Good thought

**Comment by bpoplauschi** [comment]
@rivera-ernesto Overall it looks good to me. Since it's been a while since you created it, we could merge it as it is and move from there. Let me know if I should merge it.

File: Source/SLLogger.h (hunk 2 above)
Comment (rivera-ernesto) [inline Source/SLLogger.h]: This is pretty complex as the levels have to be hardcoded in the libraries log calls. If someone adds log levels they won't get used in shipping libraries, unless they get redirected to different levels, but then those would only be aliases to the existing levels.

I like the idea of extensibility, we should think more about it.

**Comment by rivera-ernesto** [comment]
Yes, please merge it ;)

**Pull request merged**